library-best-practices/JLBP-6.md

```diff
@@ -0,0 +1,165 @@
+[JLBP-6] Package and artifact renaming rules
+--------------------------------------------
+
+When a library B in Java depends on another library A through the Maven
+repository system, there are two types of handles that library B needs to use to
+access classes in library A:
+
+1. The Maven coordinates of library A, following the form
+   `${groupId}:${artifactId}:${version}`, for example
+   `com.google.guava:guava:26.0-jre`. The Maven coordinates are used to pull the
+   library's jar from Maven Central. For each pair of group ID and artifact ID
+   (hereafter referenced as "Maven ID"), the user's build system (for example
+   Maven or Gradle) will make a final selection of only one version to put on
+   the classpath. The different build systems use different rules for which
+   version to select.
+2. The fully-qualified class names of the classes in library A. These classes
+   will generally share a Java package (the package as defined in `package`
+   statements, for example `package com.google.common.collect`). The classpath,
+   which is formed from Maven artifacts and possibly non-Maven sources, will be
+   searched for each fully-qualified class name at runtime. In order to simplify
+   the discussion, we will only talk about the Java package, assuming that the
+   classes of the library will be located in that package or subpackages.
+
+When breaking changes are introduced to library A between major version 1 and
+major version 2, a choice needs to be made: to rename or not rename? This
+question applies to both items listed above, the Maven ID (1) and the Java
+package (2).
+
+Recommendations:
+
+- *If* the new library surface is delivered under a new Java package, either use
+  a different Maven ID (different group ID or artifact ID) or bundle the old and
+  new packages together in the under the original Maven ID.
+- *If* the breaking change is made in-place and the Java package is kept the
+  same, use the same Maven ID (same group ID and same artifact ID).
+- There is a tradeoff: if an extremely large library (for example thousands of
```

Comment:
It might not be clear that this is a tradeoff between two options presented above. Maybe explicitly say "a tradeoff between the two options presented above," or modify the previous two bullets to be nested under a top-level bullet that says something like "When making a breaking change, take one of two approaches." (The latter would also let the final bullet become one of two top-level bullets, which may make sense, since (as it says) it's not solely about breaking changes.)

Reply:
Great idea, will do.